File: backend/engine/decision_engine.py
"""
mazha Tap — Decision Engine
Rule-based tap/no-tap logic. No ML, no database.

Logic flow:
1. Determine the effective tapping window for this grower.
2. Extract hourly weather data for that window.
3. Apply tree-age and rain-guard modifiers to thresholds.
4. Evaluate rain probability, rain amount, and humidity.
5. Return a structured recommendation with reasoning text.
"""
from datetime import datetime, date
from typing import Optional

from engine.config import (
    DEFAULT_TAP_START_HOUR,
    DEFAULT_TAP_END_HOUR,
    RAIN_PROB_BLOCK,
    RAIN_PROB_CAUTION,
    RAIN_PROB_CLEAR,
    RAIN_AMOUNT_BLOCK,
    RAIN_AMOUNT_CAUTION,
    HUMIDITY_CAUTION,
    HUMIDITY_HIGH,
    TREE_AGE_MODIFIERS,
    RAIN_GUARD_MODIFIER,
    LARGE_PLANTATION_LEAD_MINUTES,
    MEDIUM_PLANTATION_THRESHOLD,
)
from engine.models import (
    DecisionRequest,
    DecisionResponse,
    HourlyPoint,
    PlantationProfile,
    RecommendationLevel,
    TappingSystem,
    TreeAge,
    WindowSuggestion,
)


def _apply_modifiers(base_threshold: float, profile: PlantationProfile) -> float:
    """Apply plantation-specific multipliers to a base threshold."""
    modifier = TREE_AGE_MODIFIERS.get(profile.tree_age.value, 1.0)
    if profile.tapping_system == TappingSystem.rain_guard:
        modifier *= RAIN_GUARD_MODIFIER
    return base_threshold * modifier


def _get_tapping_window(profile: PlantationProfile) -> tuple[int, int]:
    """
    Return (start_hour, end_hour) for the effective tapping window.
    Large plantations get an earlier start time recommendation.
    """
    start = profile.tap_start_hour if profile.tap_start_hour is not None else DEFAULT_TAP_START_HOUR
    end = DEFAULT_TAP_END_HOUR

    # Large plantation lead-time: shift window earlier
    if profile.num_trees and profile.num_trees > MEDIUM_PLANTATION_THRESHOLD:
        lead_hours = LARGE_PLANTATION_LEAD_MINUTES // 60
        start = max(0, start - lead_hours)

    return start, end


def _filter_window(hourly: list[HourlyPoint], start_hour: int, end_hour: int, target_date: Optional[date] = None) -> list[HourlyPoint]:
    """Filter hourly data to only the points within the tapping window on target_date."""
    result = []
    for pt in hourly:
        try:
            dt = datetime.fromisoformat(pt.time)
        except ValueError:
            continue
        if target_date and dt.date() != target_date:
            continue
        if start_hour <= dt.hour < end_hour:
            result.append(pt)
    return result


def _find_next_safe_window(
    hourly: list[HourlyPoint],
    effective_block_prob: float,
    effective_block_rain: float,
) -> Optional[WindowSuggestion]:
    """
    Scan the next 48h for the first 3-consecutive-hour window
    where rain probability stays below the caution threshold.
    """
    for i in range(len(hourly) - 2):
        window = hourly[i:i+3]
        max_prob = max(p.precipitation_probability for p in window)
        total_rain = sum(p.rain for p in window)
        if max_prob < RAIN_PROB_CAUTION and total_rain < RAIN_AMOUNT_CAUTION:
            try:
                start_dt = datetime.fromisoformat(window[0].time)
                end_dt = datetime.fromisoformat(window[-1].time)
                confidence = "moderate" if max_prob < RAIN_PROB_CLEAR else "low"
                return WindowSuggestion(
                    start_hour=start_dt.hour,
                    end_hour=end_dt.hour + 1,
                    confidence=confidence,
                    note=f"Rain probability stays below {int(max_prob)}% — {start_dt.strftime('%A %d %b, %-I %p')} to {(end_dt.hour+1):02d}:00",
                )
            except Exception:
                pass
    return None


def evaluate(request: DecisionRequest) -> DecisionResponse:
    """Main entry point: evaluate a tap/no-tap decision."""
    profile = request.plantation
    hourly = request.hourly_forecast

    start_hour, end_hour = _get_tapping_window(profile)

    # Use today's date from the first forecast point
    target_date = None
    if hourly:
        try:
            target_date = datetime.fromisoformat(hourly[0].time).date()
        except ValueError:
            pass

    window_points = _filter_window(hourly, start_hour, end_hour, target_date)

    # Apply modifiers to thresholds
    effective_block_prob = _apply_modifiers(RAIN_PROB_BLOCK, profile)
    effective_caution_prob = _apply_modifiers(RAIN_PROB_CAUTION, profile)
    effective_block_rain = _apply_modifiers(RAIN_AMOUNT_BLOCK, profile)
    effective_caution_rain = _apply_modifiers(RAIN_AMOUNT_CAUTION, profile)

    reasoning: list[str] = []
    flags: list[str] = []  # internal flag list: "block", "caution", "humidity_high"

    # --- Weather stats for the window ---
    if window_points:
        max_rain_prob = max(p.precipitation_probability for p in window_points)
        total_rain = sum(p.rain for p in window_points)
        avg_humidity = sum(p.relative_humidity_2m for p in window_points) / len(window_points)
        max_humidity = max(p.relative_humidity_2m for p in window_points)
        peak_hour = None
        for p in window_points:
            if p.precipitation_probability == max_rain_prob:
                try:
                    peak_hour = datetime.fromisoformat(p.time).hour
                except ValueError:
                    pass
                break
    else:
        # No data for window — no forecast available; default to caution
        max_rain_prob = 0
        total_rain = 0.0
        avg_humidity = 0
        max_humidity = 0
        peak_hour = None
        reasoning.append("No forecast data available for your tapping window — proceeding with caution.")
        flags.append("caution")

    # --- Evaluate rain probability ---
    if max_rain_prob >= effective_block_prob:
        flags.append("block")
        hour_str = f" (peaking around {peak_hour:02d}:00)" if peak_hour is not None else ""
        reasoning.append(f"Rain probability reaches {max_rain_prob}% during your tapping window{hour_str} — above the safe threshold of {int(effective_block_prob)}%.")
    elif max_rain_prob >= effective_caution_prob:
        flags.append("caution")
        hour_str = f" (peaking around {peak_hour:02d}:00)" if peak_hour is not None else ""
        reasoning.append(f"Rain probability is {max_rain_prob}% during your window{hour_str} — moderate risk. Consider monitoring closely.")
    else:
        reasoning.append(f"Rain probability stays at {max_rain_prob}% during your tapping window — within safe range.")

    # --- Evaluate rain amount ---
    if total_rain >= effective_block_rain:
        if "block" not in flags:
            flags.append("block")
        reasoning.append(f"Expected rainfall of {total_rain:.1f} mm during the window exceeds safe limits ({effective_block_rain:.1f} mm).")
    elif total_rain >= effective_caution_rain:
        if "block" not in flags and "caution" not in flags:
            flags.append("caution")
        reasoning.append(f"Light rain expected ({total_rain:.1f} mm) — bark moisture may affect latex flow.")

    # --- Evaluate humidity ---
    if max_humidity >= HUMIDITY_HIGH:
        reasoning.append(f"Very high humidity ({max_humidity}%) — latex may be diluted even without direct rain.")
        if "block" not in flags:
            flags.append("caution")
    elif avg_humidity >= HUMIDITY_CAUTION:
        reasoning.append(f"High ambient humidity ({avg_humidity:.0f}% avg) — secondary risk factor noted.")

    # --- Tree age context ---
    if profile.tree_age == TreeAge.young:
        reasoning.append("Young trees are more sensitive to bark moisture — thresholds tightened accordingly.")
    elif profile.tree_age == TreeAge.old:
        reasoning.append("Older trees have slightly more bark tolerance — thresholds relaxed slightly.")

    # --- Rain-guard note ---
    if profile.tapping_system == TappingSystem.rain_guard:
        reasoning.append("Rain-guard system installed — risk thresholds relaxed by 25%.")

    # --- Large plantation note ---
    if profile.num_trees and profile.num_trees > MEDIUM_PLANTATION_THRESHOLD:
        reasoning.append(f"Large plantation ({profile.num_trees} trees) — recommended earlier start to complete tapping before rain arrives.")

    # --- Determine final recommendation & confidence ---
    if "block" in flags:
        recommendation = RecommendationLevel.dont_tap
        # Find whether any later window looks better
        next_window = _find_next_safe_window(hourly, effective_block_prob, effective_block_rain)
        if next_window:
            recommendation = RecommendationLevel.delay
            headline = f"Delay tapping — safer window available {next_window.note}"
            confidence = max(30, 80 - max_rain_prob)
        else:
            headline = "Don't tap today — rain risk is too high during your tapping window."
            confidence = max(40, 90 - max_rain_prob)
            next_window = None
    elif "caution" in flags:
        recommendation = RecommendationLevel.delay
        next_window = _find_next_safe_window(hourly, effective_block_prob, effective_block_rain)
        if next_window:
            headline = f"Tap with caution — or wait for a safer window: {next_window.note}"
        else:
            headline = "Tap with caution — some rain risk during your tapping window."
        confidence = max(40, 70 - max_rain_prob)
    else:
        recommendation = RecommendationLevel.tap
        next_window = None
        headline = "Good to tap — low rain risk during your tapping window."
        confidence = min(95, 95 - max_rain_prob)

    # Weather summary dict
    weather_summary = {
        "tapping_window": f"{start_hour:02d}:00 – {end_hour:02d}:00",
        "max_rain_probability_pct": max_rain_prob,
        "expected_rain_mm": round(total_rain, 2),
        "avg_humidity_pct": round(avg_humidity, 1),
        "max_humidity_pct": max_humidity,
    }

    return DecisionResponse(
        recommendation=recommendation,
        confidence=confidence,
        headline=headline,
        reasoning=reasoning,
        next_window=next_window,
        weather_summary=weather_summary,
    )

File: frontend/lib/engine/decision-engine.ts
import {
  DEFAULT_TAP_END_HOUR,
  DEFAULT_TAP_START_HOUR,
  HUMIDITY_CAUTION,
  HUMIDITY_HIGH,
  LARGE_PLANTATION_LEAD_MINUTES,
  LITRES_PER_BLOCK_PER_DAY,
  MEDIUM_PLANTATION_THRESHOLD,
  MIN_TREES_FOR_ONE_TAPPER,
  RAIN_AMOUNT_BLOCK,
  RAIN_AMOUNT_CAUTION,
  RAIN_GUARD_MODIFIER,
  RAIN_PROB_BLOCK,
  RAIN_PROB_CAUTION,
  RAIN_PROB_CLEAR,
  TAPPING_SEASON_MONTHS,
  TREES_PER_ACRE,
  TREES_PER_TAPPER_BLOCK,
  TREE_AGE_MODIFIERS,
} from "./config";

export type TreeAge = "young" | "mature" | "old";
export type TappingSystem = "daily" | "alternate_day" | "rain_guard" | "other";
export type LatexSaleMethod = "liquid_latex" | "rubber_sheets";
export type RecommendationLevel = "tap" | "dont_tap" | "delay";

export interface PlantationProfile {
  latitude: number;
  longitude: number;
  size_hectares?: number | null;
  num_trees?: number | null;
  tree_age: TreeAge;
  tapping_system?: TappingSystem | null;
  tap_start_hour?: number | null;
  latex_sale_method?: LatexSaleMethod | null;
}

export interface HourlyPoint {
  time: string;
  precipitation_probability: number;
  precipitation: number;
  relative_humidity_2m: number;
  temperature_2m: number;
  rain: number;
  weathercode: number;
}

export interface DecisionRequest {
  plantation: PlantationProfile;
  hourly_forecast: HourlyPoint[];
}

export interface WindowSuggestion {
  start_hour: number;
  end_hour: number;
  confidence: string;
  note: string;
}

export interface YieldEstimate {
  num_blocks: number | null;
  estimated_litres: number | null;
  tappers_needed: number | null;
  size_acres: number | null;
  note: string | null;
  off_season: boolean;
  off_season_note: string | null;
}

export interface DecisionResponse {
  recommendation: RecommendationLevel;
  confidence: number;
  headline: string;
  reasoning: string[];
  next_window: WindowSuggestion | null;
  weather_summary: {
    tapping_window: string;
    max_rain_probability_pct: number;
    expected_rain_mm: number;
    avg_humidity_pct: number;
    max_humidity_pct: number;
  };
  yield_estimate: YieldEstimate | null;
}

interface ParsedForecastTime {
  dateKey: string;
  month: number;
  hour: number;
}

function parseForecastTime(time: string): ParsedForecastTime | null {
  const match = /^(\d{4})-(\d{2})-(\d{2})T(\d{2})/.exec(time);
  if (!match) return null;

  return {
    dateKey: `${match[1]}-${match[2]}-${match[3]}`,
    month: Number(match[2]),
    hour: Number(match[4]),
  };
}

function formatWindowStart(time: string): string {
  const parsed = parseForecastTime(time);
  if (!parsed) return time;

  const date = new Date(`${parsed.dateKey}T00:00:00Z`);
  const weekday = new Intl.DateTimeFormat("en", {
    weekday: "long",
    timeZone: "UTC",
  }).format(date);
  const month = new Intl.DateTimeFormat("en", {
    month: "short",
    timeZone: "UTC",
  }).format(date);
  const day = Number(parsed.dateKey.slice(8, 10)).toString().padStart(2, "0");
  const hour12 = parsed.hour % 12 || 12;
  const period = parsed.hour < 12 ? "AM" : "PM";

  return `${weekday} ${day} ${month}, ${hour12} ${period}`;
}

function roundHalfToEven(value: number): number {
  const floor = Math.floor(value);
  const difference = value - floor;

  if (difference < 0.5) return floor;
  if (difference > 0.5) return floor + 1;
  return floor % 2 === 0 ? floor : floor + 1;
}

function roundTo(value: number, digits: number): number {
  const factor = 10 ** digits;
  return Math.round(value * factor) / factor;
}

function applyModifiers(baseThreshold: number, profile: PlantationProfile): number {
  const modifier =
    TREE_AGE_MODIFIERS[profile.tree_age] ?? TREE_AGE_MODIFIERS.mature;

  if (profile.tapping_system === "rain_guard") {
    return baseThreshold * modifier * RAIN_GUARD_MODIFIER;
  }

  return baseThreshold * modifier;
}

function getTappingWindow(profile: PlantationProfile): [number, number] {
  let start = profile.tap_start_hour ?? DEFAULT_TAP_START_HOUR;
  const end = DEFAULT_TAP_END_HOUR;

  if (profile.num_trees && profile.num_trees > MEDIUM_PLANTATION_THRESHOLD) {
    const leadHours = Math.floor(LARGE_PLANTATION_LEAD_MINUTES / 60);
    start = Math.max(0, start - leadHours);
  }

  return [start, end];
}

function filterWindow(
  hourly: HourlyPoint[],
  startHour: number,
  endHour: number,
  targetDateKey: string | null
): HourlyPoint[] {
  return hourly.filter((point) => {
    const parsed = parseForecastTime(point.time);
    if (!parsed) return false;
    if (targetDateKey && parsed.dateKey !== targetDateKey) return false;
    return startHour <= parsed.hour && parsed.hour < endHour;
  });
}

function findNextSafeWindow(hourly: HourlyPoint[]): WindowSuggestion | null {
  for (let index = 0; index <= hourly.length - 3; index += 1) {
    const window = hourly.slice(index, index + 3);
    const maxProb = Math.max(
      ...window.map((point) => point.precipitation_probability)
    );
    const totalRain = window.reduce((sum, point) => sum + point.rain, 0);

    if (maxProb < RAIN_PROB_CAUTION && totalRain < RAIN_AMOUNT_CAUTION) {
      const start = parseForecastTime(window[0].time);
      const end = parseForecastTime(window[window.length - 1].time);
      if (!start || !end) continue;

      return {
        start_hour: start.hour,
        end_hour: end.hour + 1,
        confidence: maxProb < RAIN_PROB_CLEAR ? "moderate" : "low",
        note: `Rain probability stays below ${Math.trunc(maxProb)}% — ${formatWindowStart(window[0].time)} to ${(end.hour + 1).toString().padStart(2, "0")}:00`,
      };
    }
  }

  return null;
}

function computeYieldEstimate(
  profile: PlantationProfile,
  forecastMonth: number | null
): YieldEstimate {
  const offSeason = forecastMonth
    ? !TAPPING_SEASON_MONTHS.has(forecastMonth)
    : false;
  const offSeasonNote = offSeason
    ? "You are currently outside the Kerala rubber tapping season (June–February). Tapping typically resumes in June when the monsoon begins and the trees recover from the hot dry period."
    : null;

  if (!profile.num_trees) {
    return {
      num_blocks: null,
      estimated_litres: null,
      tappers_needed: null,
      size_acres: null,
      note: null,
      off_season: offSeason,
      off_season_note: offSeasonNote,
    };
  }

  const numTrees = profile.num_trees;
  const numBlocks = Math.max(
    1,
    roundHalfToEven(numTrees / TREES_PER_TAPPER_BLOCK)
  );
  const estimatedLitres = numBlocks * LITRES_PER_BLOCK_PER_DAY;
  const tappersNeeded = numBlocks;
  const sizeAcres = roundTo(numTrees / TREES_PER_ACRE, 1);

  const parts = [
    `Your ${numTrees.toLocaleString("en-US")} trees form approx. ${numBlocks} tapper block${numBlocks !== 1 ? "s" : ""} (${TREES_PER_TAPPER_BLOCK} trees each).`,
    `Expected yield on a good tapping day: ~${Math.trunc(estimatedLitres)} litres of latex (${LITRES_PER_BLOCK_PER_DAY} L/block).`,
  ];

  if (tappersNeeded === 1) {
    parts.push("One tapper can manage your full plantation per day.");
  } else {
    parts.push(
      `${tappersNeeded} tappers would be needed to cover all blocks in one day.`
    );
  }

  if (numTrees < MIN_TREES_FOR_ONE_TAPPER) {
    parts.push(
      `With fewer than ${MIN_TREES_FOR_ONE_TAPPER} trees (${Math.floor(MIN_TREES_FOR_ONE_TAPPER / TREES_PER_ACRE)} acres), most small farmers tap their own land without hiring a tapper.`
    );
  }

  if (profile.latex_sale_method === "liquid_latex") {
    parts.push(
      "Selling as liquid latex — add acid to the collection drum to prevent premature solidification. A lab test will determine the Dry Rubber Content (DRC) for pricing."
    );
  } else if (profile.latex_sale_method === "rubber_sheets") {
    parts.push(
      "Processing into rubber sheets — pour latex into trays, add coagulating acid, pass through rollers, then sun-dry before smoke-curing. Wet weather delays the drying step."
    );
  }

  return {
    num_blocks: numBlocks,
    estimated_litres: estimatedLitres,
    tappers_needed: tappersNeeded,
    size_acres: sizeAcres,
    note: parts.join(" "),
    off_season: offSeason,
    off_season_note: offSeasonNote,
  };
}

export function evaluateDecision(request: DecisionRequest): DecisionResponse {
  const profile = request.plantation;
  const hourly = request.hourly_forecast;
  const [startHour, endHour] = getTappingWindow(profile);
  const firstForecastTime = hourly[0] ? parseForecastTime(hourly[0].time) : null;
  const targetDateKey = firstForecastTime?.dateKey ?? null;
  const yieldEstimate = computeYieldEstimate(
    profile,
    firstForecastTime?.month ?? null
  );
  const windowPoints = filterWindow(hourly, startHour, endHour, targetDateKey);
  const effectiveBlockProb = applyModifiers(RAIN_PROB_BLOCK, profile);
  const effectiveCautionProb = applyModifiers(RAIN_PROB_CAUTION, profile);
  const effectiveBlockRain = applyModifiers(RAIN_AMOUNT_BLOCK, profile);
  const effectiveCautionRain = applyModifiers(RAIN_AMOUNT_CAUTION, profile);
  const reasoning: string[] = [];
  const flags: string[] = [];
  let maxRainProb = 0;
  let totalRain = 0;
  let avgHumidity = 0;
  let maxHumidity = 0;
  let peakHour: number | null = null;

  if (yieldEstimate.off_season && yieldEstimate.off_season_note) {
    reasoning.push(`Off-season notice: ${yieldEstimate.off_season_note}`);
    flags.push("caution");
  }

  if (windowPoints.length > 0) {
    maxRainProb = Math.max(
      ...windowPoints.map((point) => point.precipitation_probability)
    );
    totalRain = windowPoints.reduce((sum, point) => sum + point.rain, 0);
    avgHumidity =
      windowPoints.reduce(
        (sum, point) => sum + point.relative_humidity_2m,
        0
      ) / windowPoints.length;
    maxHumidity = Math.max(
      ...windowPoints.map((point) => point.relative_humidity_2m)
    );
    const peakPoint = windowPoints.find(
      (point) => point.precipitation_probability === maxRainProb
    );
    peakHour = peakPoint ? parseForecastTime(peakPoint.time)?.hour ?? null : null;
  } else {
    reasoning.push(
      "No forecast data available for your tapping window — proceeding with caution."
    );
    flags.push("caution");
  }

  if (maxRainProb >= effectiveBlockProb) {
    flags.push("block");
    const hourStr = peakHour !== null ? ` (peaking around ${peakHour.toString().padStart(2, "0")}:00)` : "";
    reasoning.push(
      `Rain probability reaches ${maxRainProb}% during your tapping window${hourStr} — above the safe threshold of ${Math.trunc(effectiveBlockProb)}%.`
    );
  } else if (maxRainProb >= effectiveCautionProb) {
    flags.push("caution");
    const hourStr = peakHour !== null ? ` (peaking around ${peakHour.toString().padStart(2, "0")}:00)` : "";
    reasoning.push(
      `Rain probability is ${maxRainProb}% during your window${hourStr} — moderate risk. Consider monitoring closely.`
    );
  } else {
    reasoning.push(
      `Rain probability stays at ${maxRainProb}% during your tapping window — within safe range.`
    );
  }

  if (totalRain >= effectiveBlockRain) {
    if (!flags.includes("block")) flags.push("block");
    reasoning.push(
      `Expected rainfall of ${totalRain.toFixed(1)} mm during the window exceeds safe limits (${effectiveBlockRain.toFixed(1)} mm).`
    );
  } else if (totalRain >= effectiveCautionRain) {
    if (!flags.includes("block") && !flags.includes("caution")) {
      flags.push("caution");
    }
    reasoning.push(
      `Light rain expected (${totalRain.toFixed(1)} mm) — bark moisture may affect latex flow.`
    );
  }

  if (maxHumidity >= HUMIDITY_HIGH) {
    reasoning.push(
      `Very high humidity (${maxHumidity}%) — latex may be diluted even without direct rain.`
    );
    if (!flags.includes("block")) flags.push("caution");
  } else if (avgHumidity >= HUMIDITY_CAUTION) {
    reasoning.push(
      `High ambient humidity (${avgHumidity.toFixed(0)}% avg) — secondary risk factor noted.`
    );
  }

  if (profile.tree_age === "young") {
    reasoning.push(
      "Young trees are more sensitive to bark moisture — thresholds tightened accordingly."
    );
  } else if (profile.tree_age === "old") {
    reasoning.push(
      "Older trees have slightly more bark tolerance — thresholds relaxed slightly."
    );
  }

  if (profile.tapping_system === "rain_guard") {
    reasoning.push("Rain-guard system installed — risk thresholds relaxed by 25%.");
  }

  if (profile.num_trees && profile.num_trees > MEDIUM_PLANTATION_THRESHOLD) {
    reasoning.push(
      `Large plantation (${profile.num_trees} trees) — recommended earlier start to complete tapping before rain arrives.`
    );
  }

  let recommendation: RecommendationLevel;
  let headline: string;
  let confidence: number;
  let nextWindow: WindowSuggestion | null;

  if (flags.includes("block")) {
    recommendation = "dont_tap";
    nextWindow = findNextSafeWindow(hourly);
    if (nextWindow) {
      recommendation = "delay";
      headline = `Delay tapping — safer window available ${nextWindow.note}`;
      confidence = Math.max(30, 80 - maxRainProb);
    } else {
      headline = "Don't tap today — rain risk is too high during your tapping window.";
      confidence = Math.max(40, 90 - maxRainProb);
      nextWindow = null;
    }
  } else if (flags.includes("caution")) {
    recommendation = "delay";
    nextWindow = findNextSafeWindow(hourly);
    headline = nextWindow
      ? `Tap with caution — or wait for a safer window: ${nextWindow.note}`
      : "Tap with caution — some rain risk during your tapping window.";
    confidence = Math.max(40, 70 - maxRainProb);
  } else {
    recommendation = "tap";
    nextWindow = null;
    headline = "Good to tap — low rain risk during your tapping window.";
    confidence = Math.min(95, 95 - maxRainProb);
  }

  return {
    recommendation,
    confidence,
    headline,
    reasoning,
    next_window: nextWindow,
    weather_summary: {
      tapping_window: `${startHour.toString().padStart(2, "0")}:00 – ${endHour.toString().padStart(2, "0")}:00`,
      max_rain_probability_pct: maxRainProb,
      expected_rain_mm: roundTo(totalRain, 2),
      avg_humidity_pct: roundTo(avgHumidity, 1),
      max_humidity_pct: maxHumidity,
    },
    yield_estimate: yieldEstimate,
  };
}
